File: frontend/src/types/game.ts
import type { Player } from "./player";

export type GameStatus = "pending" | "running" | "completed" | "discarded";
export type WinnerFaction = "villagers" | "werewolves" | null;

export interface Game {
  id: string;
  created_at: string;
  status: GameStatus;
  winner: string | null;
  rounds_played: number;
  total_turns: number;
  is_degraded: boolean;
  config: Record<string, unknown> | null;
  players: Player[];
}

export interface GameListFilters {
  status?: GameStatus;
  winner?: string;
  is_degraded?: boolean;
}

File: backend/app/schemas/game.py
from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.player import PlayerResponse


class GameCreate(BaseModel):
    """Request body for creating a new game."""

    config: dict | None = Field(default=None, description="Optional game configuration")


class GameResponse(BaseModel):
    """Response for a single game (list view)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    status: str
    winner: str | None = None
    rounds_played: int
    total_turns: int
    is_degraded: bool
    config: dict | None = None


class GameStateResponse(BaseModel):
    """Full game state including players (detail view)."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    status: str
    winner: str | None = None
    rounds_played: int
    total_turns: int
    is_degraded: bool
    config: dict | None = None
    players: list[PlayerResponse] = []
